### PR TITLE
feat: add bookmark folder page

### PR DESCRIPTION
**Pull request by ilasw**
## Changes
- Add read it later bookmark folder
- Add folder single page
- Some refactor in Bookmark-related feed components 

## Events

Did you introduce any new tracking events?

## Experiment

Did you introduce any new experiments?

## Manual Testing

> [!CAUTION]
> Please make sure existing components are not breaking/affected by this PR

### Preview domain
https://mi-669-fe-add-new-bookmark-folde.preview.app.daily.dev

### CODE REVIEW

File: packages/shared/src/graphql/bookmarks.ts [hunk 1]
```diff
@@ -89,6 +109,45 @@ export const createBookmarkFolder = async ({
...
+    }
+  }
+`;
+
+export const updateBookmarkFolder = async ({
+  id,
+  name,
+  icon,
+}: BookmarkFolder): Promise<BookmarkFolder> => {
+  return gqlClient
```
Comment (sshanzel) [inline packages/shared/src/graphql/bookmarks.ts:126]: Probably the reason why `onError` is not working as intended. We should probably use try-catch.

File: packages/shared/src/components/bookmark/BookmarkFolderContextMenu.tsx [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    const confirm = await showPrompt({
+      title: `Delete ${folder.name}?`,
+      description: 'This will also delete all bookmarks in this folder.',
+      cancelButton: { title: 'Cancel' },
+      okButton: { title: 'Delete folder' },
+    });
+
+    if (confirm) {
+      await router.push('/bookmarks');
+      await deleteFolder.mutate(folder.id);
```
Comment (sshanzel) [inline packages/shared/src/components/bookmark/BookmarkFolderContextMenu.tsx:40]: Should it probably be the other way around?
```suggestion
      await deleteFolder.mutate(folder.id);
      await router.push('/bookmarks');
```

**Review by sshanzel — APPROVED** [review]
Don't have much to add. Code-wise looked amazing 🚀

File: packages/shared/src/components/bookmark/BookmarkFolderContextMenu.tsx (hunk 2 above)
Comment (ilasw) [reply packages/shared/src/components/bookmark/BookmarkFolderContextMenu.tsx:40]: I've tried, but page glitch too much because we have an empty page at start (probably due to optimistic update) then redirect 🤔

**Review by AmarTrebinjac — CHANGES_REQUESTED** [review]
Nice stuff, just a couple of comments! :D

File: packages/shared/src/hooks/bookmark/useBookmarkFolder.ts [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+  const { displayToast } = useToastNotification();
+  const queryClient = useQueryClient();
+  const { logEvent } = useLogContext();
+
+  const {
+    isPending: isPendingQuery,
+    isSuccess,
+    folders,
+  } = useBookmarkFolderList();
+  const folder = useMemo(() => folders.find((f) => f.id === id), [folders, id]);
```
[the remark this answers is not in the document]
Comment (capJavert) [reply packages/shared/src/hooks/bookmark/useBookmarkFolder.ts:45]: it think its good rule of the thumb to memoize iterations, since folders can grow in size and no point in running each render since its blocking operation

File: packages/webapp/pages/bookmarks/[folderId].tsx [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+const seo: NextSeoProps = {
+  title: `Your daily.dev bookmarks`,
+  openGraph: { ...defaultOpenGraph },
+  ...defaultSeo,
+};
+
+const BookmarksPage = (): ReactElement => {
+  const router = useRouter();
+  const id = `${router.query.folderId}`;
```
[the remark this answers is not in the document]
Comment (ilasw) [reply packages/webapp/pages/bookmarks/[folderId].tsx:19]: Because values from `router.query` can be also string[] so we need to fix it for ts checks

File: packages/webapp/pages/bookmarks/[folderId].tsx [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+  const layout = getBookmarkFeedLayout(null, { seo }, props);
+
+  if (!folder) {
+    if (isReady && !isPending) {
+      router.replace('/bookmarks');
+    }
+
+    return null;
+  }
```
[the remark this answers is not in the document]
Comment (ilasw) [reply packages/webapp/pages/bookmarks/[folderId].tsx:41]: Thanks, updated!

File: packages/webapp/pages/bookmarks/[folderId].tsx [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+  const id = `${router.query.folderId}`;
+  const {
+    query: { folder, isPending, isReady },
+  } = useBookmarkFolder({ id });
+  const props = useMemo(() => {
+    return {
+      ...bookmarkFeedLayoutProps,
+      ...(folder && {
+        folder,
+        title: `${folder.icon} ${folder.name}`,
```
[the remark this answers is not in the document]
Comment (ilasw) [reply packages/webapp/pages/bookmarks/[folderId].tsx]: Thanks, added ✅

File: packages/webapp/pages/bookmarks/[folderId].tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (capJavert) [reply packages/webapp/pages/bookmarks/[folderId].tsx:19]: Ah, yeah its fine

File: packages/shared/src/hooks/bookmark/useBookmarkFolder.ts [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+    onSuccess: (updated) => {
+      displayToast(`Folder ${updated.name} updated`);
+
+      logEvent({
+        event_name: LogEvent.RenameBookmarkFolder,
+        target_id: folder.id,
+      });
+
+      const listQueryKey = generateQueryKey(RequestKey.BookmarkFolders);
+      queryClient.setQueryData(listQueryKey, (data: BookmarkFolder[]) => {
```
[the remark this answers is not in the document]
Comment (ilasw) [reply packages/shared/src/hooks/bookmark/useBookmarkFolder.ts]: Nice, updated using `onMutate` ✔️

File: packages/shared/src/hooks/bookmark/useBookmarkFolder.ts [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+    onSuccess: () => {
+      displayToast(`Folder deleted`);
+
+      logEvent({
+        event_name: LogEvent.DeleteBookmarkFolder,
+        target_id: folder.id,
+      });
+
+      const listQueryKey = generateQueryKey(RequestKey.BookmarkFolders);
+      queryClient.setQueryData(listQueryKey, (data: BookmarkFolder[]) => {
```
[the remark this answers is not in the document]
Comment (ilasw) [reply packages/shared/src/hooks/bookmark/useBookmarkFolder.ts]: Nice, updated using `onMutate` ✔️

**Review by rebelchris — CHANGES_REQUESTED** [review]
Nice, some minor questions

File: packages/webapp/pages/bookmarks/[folderId].tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (rebelchris) [reply packages/webapp/pages/bookmarks/[folderId].tsx:19]: Ah you can also use `as string` btw :)

File: packages/shared/src/components/BookmarkFeedLayout.tsx [hunk 9]
```diff
@@ -73,33 +102,32 @@ export default function BookmarkFeedLayout({
...
     };
-  }, [defaultKey, searchQuery]);
+  }, [searchQuery, defaultKey, listId, isReminderOnly]);
 
-  if (showEmptyScreen) {
+  useEffect(() => {
+    if (folder?.id) {
+      setShowEmptyScreen(false);
+    }
+  }, [folder?.id]);
```
[the remark this answers is not in the document]
Comment (ilasw) [reply packages/shared/src/components/BookmarkFeedLayout.tsx:118]: Moving from one folder to another we need to reset the empty screen state (tested and was odd to have empty screen and results together)

[the remark this answers is not in the document]
Comment (rebelchris) [reply packages/shared/src/components/BookmarkFeedLayout.tsx:118]: Ah but shouldn't it be based on empty result to render emptyFeed or not?

File: packages/shared/src/components/BookmarkFeedLayout.tsx [hunk 10]
```diff
@@ -128,6 +166,7 @@ export default function BookmarkFeedLayout({
           />
         </div>
       )}
+      {isFolderPage && showEmptyScreen && <BookmarkEmptyScreen />}
```
[the remark this answers is not in the document]
Comment (ilasw) [reply packages/shared/src/components/BookmarkFeedLayout.tsx:169]: This is only for folders since we have a different layout but I think we could uniform it to the folder one. Wdyt?

[the remark this answers is not in the document]
Comment (rebelchris) [reply packages/shared/src/components/BookmarkFeedLayout.tsx:169]: Ah yes, should be able to uniform it I guess since it's the same screen.

File: packages/webapp/pages/bookmarks/later.tsx [hunk 11]
```diff
@@ -0,0 +1,24 @@
+import React, { ReactElement } from 'react';
+import { NextSeoProps } from 'next-seo/lib/types';
+import {
+  getBookmarkFeedLayout,
+  bookmarkFeedLayoutProps,
+} from '../../components/layouts/BookmarkFeedPage';
+import { defaultOpenGraph, defaultSeo } from '../../next-seo';
+
+const seo: NextSeoProps = {
+  title: `Your daily.dev read it later bookmarks`,
```
[the remark this answers is not in the document]
Comment (sshanzel) [reply packages/webapp/pages/bookmarks/later.tsx:10]: They decided to not focus on this yet since it is currently a private endpoint.

File: packages/shared/src/components/BookmarkFeedLayout.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (ilasw) [reply packages/shared/src/components/BookmarkFeedLayout.tsx:118]: Yes, but seems like there is no way to reset it from the inside, I'm looking if I missed some param 🤔

File: packages/shared/src/components/BookmarkFeedLayout.tsx [hunk 12]
```diff
@@ -73,33 +102,32 @@ export default function BookmarkFeedLayout({
...
-  if (showEmptyScreen) {
+  useEffect(() => {
+    if (folder?.id) {
+      setShowEmptyScreen(false);
+    }
+  }, [folder?.id]);
+
+  if (showEmptyScreen && !isFolderPage) {
     return <BookmarkEmptyScreen />;
   }
```
[the remark this answers is not in the document]
Comment (ilasw) [reply packages/shared/src/components/BookmarkFeedLayout.tsx]: Updated ✔️

File: packages/shared/src/components/BookmarkFeedLayout.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (ilasw) [reply packages/shared/src/components/BookmarkFeedLayout.tsx]: Using `emptyScreen` feedProp solved ✅

File: packages/shared/src/components/bookmark/BookmarkFolderContextMenu.tsx [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+  const { showPrompt } = usePrompt();
+  const { update: updateFolder, delete: deleteFolder } = useBookmarkFolder({
+    id: folder?.id,
+  });
+
+  const handleDelete = async () => {
+    const confirm = await showPrompt({
+      title: `Delete ${folder.name}?`,
+      description: 'This will also delete all bookmarks in this folder.',
+      cancelButton: { title: 'Cancel' },
```
[the remark this answers is not in the document]
Comment (ilasw) [reply packages/shared/src/components/bookmark/BookmarkFolderContextMenu.tsx:34]: Removed ✔️

File: packages/shared/src/components/bookmark/BookmarkFolderContextMenu.tsx [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+      okButton: { title: 'Delete folder' },
+    });
+
+    if (confirm) {
+      await router.push('/bookmarks');
```
[the remark this answers is not in the document]
Comment (ilasw) [reply packages/shared/src/components/bookmark/BookmarkFolderContextMenu.tsx:39]: This was also pointer in a prev comment, the issue is that deleting the folder and only after move the user to another page cause glitches, do we want to do it?

File: packages/shared/src/hooks/bookmark/useBookmarkFolder.ts [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+import { useBookmarkFolderList } from './useBookmarkFolderList';
+import { useToastNotification } from '../useToastNotification';
+import { generateQueryKey, RequestKey } from '../../lib/query';
+import { EmptyResponse } from '../../graphql/emptyResponse';
+import { useLogContext } from '../../contexts/LogContext';
+import { LogEvent } from '../../lib/log';
+
+type UseBookmarkFoldersProps = Pick<BookmarkFolder, 'id'>;
+
+interface UseBookmarkFolder {
```
[the remark this answers is not in the document]
Comment (ilasw) [reply packages/shared/src/hooks/bookmark/useBookmarkFolder.ts:17]: I can split the hook in 3 but they are used all in the same page at once (context menu). So i preferred to have only in one with a well separated return object. 

Let me know If splitting is a better idea in your opinion ✔️

File: packages/shared/src/components/BookmarkFeedLayout.tsx [hunk 16]
```diff
@@ -108,8 +136,18 @@ export default function BookmarkFeedLayout({
...
         {searchChildren}
-        {shareBookmarksButton('hidden laptop:flex ml-4', 'Share bookmarks')}
-        {shareBookmarksButton('flex laptop:hidden ml-4')}
+        {!isFolderPage && (
+          <ShareBookmarksButton
+            className="ml-4 flex"
+            icon={<ShareIcon secondary={showSharedBookmarks} />}
+            onClick={() => setShowSharedBookmarks(true)}
+          >
+            <span className="hidden laptop:inline">Share bookmarks</span>
```
[the remark this answers is not in the document]
Comment (ilasw) [reply packages/shared/src/components/BookmarkFeedLayout.tsx:134]: I would like to avoid to add `iconOnly` because many times I found issues on hydration when using `useViewSize` for add/remove props, [before this button didn't have it](https://github.com/dailydotdev/apps/pull/3948/files/0036473ef814dafba417ecbb392b8f8842dece50#diff-2802a541bd3f1481e3cc19c9d15a4a746b9a8a549628bc5a4769c9c6c2ee683aL87) and it is changed only because of a refactor (before was duplicated). Let me know if action is required here ✔️

File: packages/shared/src/components/BookmarkFeedLayout.tsx [hunk 17]
```diff
@@ -73,33 +101,30 @@ export default function BookmarkFeedLayout({
       feedQueryKey: defaultKey,
       query: BOOKMARKS_FEED_QUERY,
       variables: {
+        ...(listId && { listId }),
+        reminderOnly: isReminderOnly,
         supportedTypes: supportedTypesForPrivateSources,
       },
-      onEmptyFeed: () => setShowEmptyScreen(true),
+      emptyScreen: (
```
Comment (rebelchris) [inline packages/shared/src/components/BookmarkFeedLayout.tsx:108]: :amaze:

File: packages/shared/src/components/BookmarkFeedLayout.tsx (hunk 16 above)
[the remark this answers is not in the document]
Comment (rebelchris) [reply packages/shared/src/components/BookmarkFeedLayout.tsx:142]: Oh recently fixed it to work on client side here:
https://github.com/dailydotdev/apps/commit/94bd3b9ad751916c19dd53156e4db57b39e5cda9#diff-6c7ea4387e2880c7e85248c8c15f8901aeea14bffd9b2629d1530012f87118f3R48

But up to you, not blocking if it looks ok.

File: packages/shared/src/components/bookmark/BookmarkFolderContextMenu.tsx (hunk 14 above)
[the remark this answers is not in the document]
Comment (rebelchris) [reply packages/shared/src/components/bookmark/BookmarkFolderContextMenu.tsx:38]: Maybe it shouldn't await it?
I'm just worried in some case it could redirect and not execute the deleteFolder at all? (is it possible?)

**Review by rebelchris — APPROVED** [review]
Nothing much blocking here, just minor comments but up to you

File: packages/webapp/pages/bookmarks/[folderId].tsx [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+    };
+  }, [folder]);
+
+  const layout = getBookmarkFeedLayout(null, { seo }, props);
+
+  useEffect(() => {
+    if (isReady && !isPending && !folder) {
+      router.replace('/bookmarks');
+    }
+  }, [folder, isPending, isReady, router]);
```
[the remark this answers is not in the document]
Comment (rebelchris) [reply packages/webapp/pages/bookmarks/[folderId].tsx:39]: We can do it with SSR yes, maybe actually not a bad idea for bookmarks they don't really benefit from ISR anyway, and probably even faster on SSR.

But not blocking we can do it later after release even?
Also love to here thoughts from @sshanzel and @capJavert on it.

File: packages/shared/src/hooks/bookmark/useBookmarkFolder.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (rebelchris) [reply packages/shared/src/hooks/bookmark/useBookmarkFolder.ts:17]: Let's leave it for now. :)

File: packages/webapp/pages/bookmarks/[folderId].tsx (hunk 18 above)
[the remark this answers is not in the document]
Comment (capJavert) [reply packages/webapp/pages/bookmarks/[folderId].tsx]: I would leave it as is for now and see where it goes after. For example if bookmark folders become share-able we will have to change some stuff either way.

**Review by capJavert — APPROVED** [review]
